### ISSUE TITLE
[bitnami/redis] sentinel failover in prestop hook causes write loss

### ISSUE THREAD

**Issue by simon-baatz** [bitnami/charts#12598]
### Name and Version

bitnami/redis 17.1.6

### What steps will reproduce the bug?

1. Start a sentinel deployment with default config:  `helm install redis
   bitnami/redis --set sentinel.enabled=true`. Wait until all nodes are up.
   Usually, `redis-node-0` will be the master. This is assumed in the following.

2. Create writes (once per second is enough), the window during which writes are
   lost is typically a couple of seconds. For example, increment a key every
   second and read it on a replica.

   ```bash
   while /bin/true ; do  echo -n "W: " ; kubectl exec redis-node-0 --container redis -- bash -c 'REDISCLI_AUTH="$REDIS_PASSWORD" redis-cli -h localhost -p 6379 incr key' ; sleep 1 ; echo -n "R: " ; kubectl exec redis-node-1 --container redis -- bash -c 'REDISCLI_AUTH="$REDIS_PASSWORD" redis-cli -h localhost -p 6379 get key' ; done
   ```

   (While this is running, collect the logs of `redis-node-0`: `kubectl logs redis-node-0 --container=redis -f` for later analysis)

3. Shutdown the master by deleting it `kubectl delete pod redis-node-0`. Kubernetes will re-create the pod (as a replica node).

4. Observe that although writes are succeeding during failover, they are lost after the failover.

### Are you using any custom parameters or values?

Enable sentinel deployment as shown above.

### What is the expected behavior?

If no error occurs for a write, the written value should not be lost (there are
cases in which Redis does not guarantee this, but a regular shutdown should not be
such a case as Redis explicitly tries to prevent that with its shutdown procedure).

### What do you see instead?

A typical output of the INCR/GET loop above looks like:

```
W: 3
R: 3
W: 4
R: 4
W: 5
R: 4
W: 6
R: 4
W: 7
R: 4
W: 8
R: 4
W: 9
R: 4
W: 10
R: 4
W: 11
R: 4
W: error: Internal error occurred: error executing command in container: container not running (c1318323ffb243cdc98658b6a9c504121422a3f71bc104932474f57fc380d811)
```

Writes to the old master succeed, but the values are not replicated anymore. The
time window during which this happens is surprisingly large. (Note that
according to the client spec, the client only needs to verify the current master
in case of a reconnection. The original problem occurred with a sentinel
protocol aware Jedis client. In this bug report, `redis-cli` is used to make the
problem easier to reproduce.)

The master node-0 loses its replica connections early, but it takes more than
ten seconds before the master is actually turned into a replica (during this
time, we have two masters):

```
1:M 19 Sep 2022 18:38:15.519 # Connection with replica redis-node-1.redis-headless.default.svc.cluster.local:6379 lost.
1:M 19 Sep 2022 18:38:16.738 # Connection with replica redis-node-2.redis-headless.default.svc.cluster.local:6379 lost.
1:S 19 Sep 2022 18:38:27.314 * Before turning into a replica, using my own master parameters to synthesize a cached master: I may be able to synchronize with the new master with just a partial transfer.
1:S 19 Sep 2022 18:38:27.314 * Connecting to MASTER redis-node-1.redis-headless.default.svc.cluster.local:6379
1:S 19 Sep 2022 18:38:27.324 * MASTER <-> REPLICA sync started
1:S 19 Sep 2022 18:38:27.325 * REPLICAOF redis-node-1.redis-headless.default.svc.cluster.local:6379 enabled (user request from 'id=92 addr=10.244.0.13:53524 laddr=10.244.2.13:6379 fd=9 name=sentinel-9fe32540-cmd age=10 idle=0 flags=x db=0 sub=0 psub=0 ssub=0 multi=4 qbuf=241 qbuf-free=20233 argv-mem=4 multi-mem=221 rbs=1024 rbp=1024 obl=45 oll=0 omem=0 tot-mem=22617 events=r cmd=exec user=default redir=-1 resp=2')
```

When node-1 becomes the new master, both replicas have newer data but it is discarded:

```
1:M 19 Sep 2022 18:38:15.521 * MASTER MODE enabled (user request from 'id=3 addr=10.244.1.18:50122 laddr=10.244.1.18:6379 fd=10 name=sentinel-33535e4e-cmd age=154 idle=0 flags=x db=0 sub=0 psub=0 ssub=0 multi=4 qbuf=188 qbuf-free=20286 argv-mem=4 multi-mem=169 rbs=2048 rbp=1024 obl=45 oll=0 omem=0 tot-mem=23589 events=r cmd=exec user=default redir=-1 resp=2')
1:M 19 Sep 2022 18:38:16.773 * Replica redis-node-2.redis-headless.default.svc.cluster.local:6379 asks for synchronization
1:M 19 Sep 2022 18:38:16.773 * Partial resynchronization not accepted: Requested offset for second ID was 47473, but I can reply up to 46784
1:M 19 Sep 2022 18:38:16.773 * Delay next BGSAVE for diskless SYNC
1:M 19 Sep 2022 18:38:21.897 * Starting BGSAVE for SYNC with target: replicas sockets
1:M 19 Sep 2022 18:38:21.898 * Background RDB transfer started by pid 738
738:C 19 Sep 2022 18:38:21.899 * Fork CoW for RDB: current 0 MB, peak 0 MB, average 0 MB
1:M 19 Sep 2022 18:38:21.900 # Diskless rdb transfer, done reading from pipe, 1 replicas still up.
1:M 19 Sep 2022 18:38:21.998 * Background RDB transfer terminated with success
1:M 19 Sep 2022 18:38:21.998 * Streamed RDB transfer with replica redis-node-2.redis-headless.default.svc.cluster.local:6379 succeeded (socket). Waiting for REPLCONF ACK from slave to enable streaming
1:M 19 Sep 2022 18:38:21.998 * Synchronization with replica redis-node-2.redis-headless.default.svc.cluster.local:6379 succeeded
1:M 19 Sep 2022 18:38:27.342 * Replica redis-node-0.redis-headless.default.svc.cluster.local:6379 asks for synchronization
1:M 19 Sep 2022 18:38:27.342 * Partial resynchronization not accepted: Requested offset for second ID was 49609, but I can reply up to 46784
```

### Additional information

MR https://github.com/bitnami/charts/pull/5528 added a pre-stop hook that
initiates a "SENTINEL FAILOVER" if a master shuts down. However, sentinel
failover is not safe to use in this situation, as the sentinel just assumes that
the current master is down. See https://github.com/redis/redis/issues/6062 for
details (note: that issue is closed because Redis implements a "FAILOVER"
command now, not because the behavior of "SENTINEL FAILOVER" changed).

Presumably, the time window for lost writes was very small (or not existent) in
most situations with this change, as the redis master got a SIGTERM at almost
the same time.

This changed with https://github.com/bitnami/charts/pull/6080, which delayed the
redis shutdown until the master became a replica.

Here are the options for a fix I can think of:

1. Use the new "FAILOVER" command instead (the "fix" for
https://github.com/redis/redis/issues/6062). I had no luck with this as this
command just switches the role of two redis servers. In a 3 node sentinel setup,
another failover initiated by sentinel always occurred in my tests.

2. Revert https://github.com/bitnami/charts/pull/6080. However, there is
probably still the possibility for a race condition here if there are lagging
replicas.

3. Implement the proposal from
https://github.com/redis/redis/issues/6062#issuecomment-489553457: Issue a
"CLIENT PAUSE" and wait for replicas to catch up before initiating a "SENTINEL
FAILOVER". It looks like this can't be done externally, as the master needs to
send an internal "REPLCONF" command to get acks quickly, see
https://github.com/redis/redis/blob/7.0.4/src/server.c#L4036.

4. As of Redis 7, "SHUTDOWN" includes a phase (up to 10 seconds) that does
exactly that (i.e. "CLIENT PAUSE WRITE" & wait for replicas to catch up) if there are
lagging replicas. As "SHUTDOWN" blocks until the server terminates, we could
issue a "SHUTDOWN" before doing the "SENTINEL FAILOVER".

   However, SHUTDOWN may take longer than just this phase. If no RDB is used, this
   shouldn't be much (flush AOF). However, if a save point needs to be written,
   this may take time. We could initiate the "SENTINEL FAILOVER" after 10 seconds
   at the latest, though (if failover has not happened already by that time).

All of these seem to have problems/edge cases of their own. It would be best to
get 1. or 3. to work. If not, maybe we could do 2. and include 4. as an option
for those who want additional safety?

**Comment by jotamartos** [comment]
Hi @simon-baatz,

Sorry for the delay here. We are going to take a look at this issue and will update it when having more information. 

Thanks

**Comment by jotamartos** [comment]
Hi @simon-baatz,

I just created this PR to solve this issue. I checked that the write requests are stopped when deleting the pod and everything works fine again either by writing in the new master server or by running the `sentinel failover mymaster` command to recover the master in the new pod. 

https://github.com/bitnami/charts/pull/13021

I'll wait for your feedback during some days and merge the changes after that. 

Thanks
